Libraries/CO.CDP.AspNetCore/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,49 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using static CO.CDP.OrganisationInformation.Persistence.IPersonRepository.PersonRepositoryException;
+using static CO.CDP.OrganisationInformation.Persistence.ITenantRepository.TenantRepositoryException;
+
+
+namespace CO.CDP.AspNetCore;
+
+public static class ServiceCollectionExtensions
+{
+    public static IServiceCollection AddCDPProblemDetails(this IServiceCollection services)
+    {
+        services.AddProblemDetails(options =>
+        {
+            options.CustomizeProblemDetails = ctx =>
+            {
+                switch (ctx.Exception)
```

Comment:
This is a slippery slope to a God Switch Statement (tm) that knows about every single package in the project.

Reply:
Let me share some red flags I noticed so far.

All the Web API will depend on this component, which makes it `stable`. Web API components are `volatile` as they will change a lot to implement the always changing business landscape.

How would we handle an exception thrown in Web API's use case service?

The Stable-Dependencies Principle (SDP) teaches us:

> Depend in the direction of stability.

Web APIs will depend on CO.CDP.AspNetCore, that's unavoidable. That makes the AspNetCore package the stable one.

As a result CO.CDP.AspNetCore should not depend on Web APIs. With the current design, it will need to.

That will also lead to cyclic dependencies, while The Acyclic Dependencies Principle (ADP) suggests:

> Allow no cycles in the component dependency graph.

Finally, using specific types in this component makes it less abstract, while the Stable-Abstractions Principle (SAP) says:

> A component should be as abstract as it is stable.

I'm quoting principles, as that's their purpose. For a detailed explanation, see [Chapter 28 of Agile Principles, Patterns, and Practices in C#](https://learning.oreilly.com/library/view/agile-principles-patterns/0131857258/ch28.xhtml) or numerous blog posts on Principles of Package and Component Design.

Reply:
What can we do to make this component a bit more abstract? Specifically, make it doesn't depend on specific exception types from other application components?